core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -1051,11 +1058,13 @@ private[spark] class BlockManager(
           }
           blockIsUpdated = true
         }
+        pendingToRemove.put(blockId, currentTaskAttemptId)
```

Comment:
This put should happen as soon as we check `containsKey` in L1035. Between now and then many things could happen; another thread may find that `containsKey` is false at the same time and both threads can still end up trying to drop the block.